controllers/openstackdataplanedeployment_controller.go

```diff
@@ -0,0 +1,245 @@
+		// Error reading the object - requeue the request.
+		return ctrl.Result{}, err
+	}
+
+	helper, _ := helper.NewHelper(
+		instance,
+		r.Client,
+		r.Kclient,
+		r.Scheme,
+		logger,
+	)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Always patch the instance status when exiting this function so we can persist any changes.
+	defer func() { // update the Ready condition based on the sub conditions
+		if instance.Status.Conditions.AllSubConditionIsTrue() {
+			instance.Status.Conditions.MarkTrue(
+				condition.ReadyCondition, condition.ReadyMessage)
+		} else {
+			// something is not ready so reset the Ready condition
+			instance.Status.Conditions.MarkUnknown(
+				condition.ReadyCondition, condition.InitReason, condition.ReadyInitMessage)
+			// and recalculate it based on the state of the rest of the conditions
+			instance.Status.Conditions.Set(
+				instance.Status.Conditions.Mirror(condition.ReadyCondition))
+		}
+		err := helper.PatchInstance(ctx, instance)
+		if err != nil {
+			logger.Error(err, "Error updating instance status conditions")
+			_err = err
+			return
+		}
+	}()
+
+	// If the deploy is already done, return immediately.
+	if instance.Status.Deployed {
+		logger.Info("Already deployed", "instance.Status.Deployed", instance.Status.Deployed)
+		return ctrl.Result{}, nil
+	}
+
+	// Initialize Status
+	if instance.Status.Conditions == nil {
+		instance.InitConditions()
+		// Register overall status immediately to have an early feedback e.g.
+		// in the cli
+		return ctrl.Result{}, nil
+	}
+
+	// All setup tasks done
+	// Mark SetupReadyCondition=True
+	instance.Status.Conditions.MarkTrue(dataplanev1.SetupReadyCondition, condition.ReadyMessage)
+
+	// Ensure NodeSets
+	nodeSets := dataplanev1.OpenStackDataPlaneNodeSetList{}
+	for _, nodeSet := range instance.Spec.NodeSets {
+
+		// Fetch the OpenStackDataPlaneNodeSet instance
+		nodeSetInstance := &dataplanev1.OpenStackDataPlaneNodeSet{}
+		err := r.Client.Get(
+			ctx,
+			types.NamespacedName{
+				Namespace: instance.GetNamespace(),
+				Name:      nodeSet},
+			nodeSetInstance)
+		if err != nil {
+			if k8s_errors.IsNotFound(err) {
+				// Request object not found, could have been deleted after reconcile request.
+				// Owned objects are automatically garbage collected.
+				// For additional cleanup logic use finalizers. Return and don't requeue.
+				return ctrl.Result{}, nil
+			}
+			// Error reading the object - requeue the request.
+			return ctrl.Result{}, err
+		}
+		nodeSets.Items = append(nodeSets.Items, *nodeSetInstance)
+	}
+
+	// All nodeSets successfully fetched.
+	// Mark InputReadyCondition=True
+	instance.Status.Conditions.MarkTrue(condition.InputReadyCondition, condition.ReadyMessage)
+
+	// Deploy each nodeSet
+	// TODO(slagle)
+	// Make this work in parallel for nodeSets. Call Deploy for each NodeSet
+	// and save the deployResult in a list. Then check the results outside this
+	// loop. That will start each nodeSet deploying in parallel.
```

Comment:
Probably a good place for Go routines and each thread can report back via a channel?

Reply:
For now, I've got the NodeSets deploying in parallel. Even though they are started and checked sequentially, they'll be running in parallel. I think this is simpler for the time being. We can look at the Go routine approach as a follow up perhaps.